windows/devices/uwb/UwbSession.cxx

```diff
@@ -178,69 +178,39 @@ UwbStatus
-    // UWB_APP_CONFIG_PARAM *appConfigParamNumberOfControlees = appConfigParamList + 0;
-    // UWB_APP_CONFIG_PARAM *appConfigParamDstMacAddress = appConfigParamList + 1;
-
-    // // Populate NUMBER_OF_CONTROLEES app configuration parameter.
-    // auto &numberOfControleesPayload = *reinterpret_cast<uint8_t *>(appConfigParamNumberOfControlees + 1);
-    // appConfigParamNumberOfControlees->paramType = UWB_APP_CONFIG_PARAM_TYPE_NUMBER_OF_CONTROLEES;
-    // appConfigParamNumberOfControlees->paramLength = 1;
-    // numberOfControleesPayload = static_cast<uint8_t>(m_peers.size());
-
-    // // Populate DST_MAC_ADDRESS app configuration parameter.
-    // auto dstMacAddressPayload = reinterpret_cast<uint8_t *>(appConfigParamDstMacAddress + 1);
-    // appConfigParamDstMacAddress->paramType = UWB_APP_CONFIG_PARAM_TYPE_DST_MAC_ADDRESS;
-    // appConfigParamDstMacAddress->paramLength = static_cast<uint32_t>(macAddressessLength);
-    // auto dstMacAddress = dstMacAddressPayload;
-    // for (const auto &peer : m_peers) {
-    //     const auto value = peer.GetValue();
-    //     std::copy(std::cbegin(value), std::cend(value), dstMacAddress);
-    //     std::advance(dstMacAddress, std::size(value));
-    // }
-
-    // // Attempt to set all new parameters.
-    // DWORD bytesReturned = 0;
-    // UWB_SET_APP_CONFIG_PARAMS_STATUS appConfigParamsStatus; // TODO: this needs to be dynamically allocated to fit returned content
-    // BOOL ioResult = DeviceIoControl(m_handleDriver.get(), IOCTL_UWB_SET_APP_CONFIG_PARAMS, &appConfigParams, static_cast<DWORD>(appConfigParamsSize), &appConfigParamsStatus, sizeof appConfigParamsStatus, &bytesReturned, nullptr);
-    // if (!LOG_IF_WIN32_BOOL_FALSE(ioResult)) {
-    //     // TODO
-    //     HRESULT hr = GetLastError();
-    //     PLOG_ERROR << "could not send params to driver, hr=" << std::showbase << std::hex << hr;
-    // }
+    if (!m_uwbSessionConnector) {
+        PLOG_WARNING << "No associated connector";
+        return UwbStatusGeneric::Failed;
+    }
+
+    auto params = GetApplicationConfigurationParameters({ UwbApplicationConfigurationParameterType::DestinationMacAddresses });
+    if (std::size(params) != 1) {
+        throw std::runtime_error("GetApplicationConfigurationParameters() for 1 parameter did not return exactly 1 result. This is a bug!");
+    }
+    auto macAddresses = std::get<std::unordered_set<::uwb::UwbMacAddress>>(params.front().Value);
+    auto [_, inserted] = macAddresses.insert(controleeMacAddress);
+    if (!inserted) {
+        PLOG_INFO << "controleeMacAddress already added, skipping";
+        return UwbStatusGeneric::Ok;
+    }
+    if (std::size(macAddresses) > MaximumNumberOfControleesInMulticastSession) {
+        PLOG_WARNING << "exceeded max number of controlees with numberOfControlees=" << std::size(macAddresses);
+        return UwbStatusGeneric::Rejected;
```

Comment:
This seems a bit too generic. Can you consult the spec and see what the expected status should be when # of controlees exceeds the max? I'm thinking either `UwbStatusSession::MulticastListFull`, `UwbStatusGeneric::InvalidRange`, or `UwbStatusGeneric::InvalidParameter`.

Reply:
From my understanding of the UCI spec, it should be `STATUS_ERROR_MULTICAST_LIST_FULL`: "Indicates when multicast list is full during one to many ranging"

Reply:
I chose InvalidRange because NumberOfControlees would be out of the range

Reply:
> From my understanding of the UCI spec, it should be `STATUS_ERROR_MULTICAST_LIST_FULL`: "Indicates when multicast list is full during one to many ranging"
> 
> InvalidParameter/Range is used for configuration parameters

I think you're right, while we are setting configuration parameters, STATUS_ERROR_MULTICAST_LIST_FULL fits better here. I had thought that it was specific to SESSION_UPDATE_CONTROLLER_MULTICAST_LIST_NTF, but it looks like that notification has its own table of statuses.

Reply:
On the other hand, you might even say that we should let the UWBS return the status, and the role of this function is just to forward that status to the caller. However, the way I implemented it right now is a bit more efficient (but it does guess what the UWBS would return).

Reply:
I'd fall back to doing what the spec says. If we ever have to get FIRA certified, then we have to do what the spec says. So, I think our options are to either return the spec value here, or, as you suggested @forwardpointer, just pass this through to the driver and let the device provide the status. I'm thinking of preferring the latter since the less the service is involved, the better.

Reply:
Thinking about this more, I realize that STATUS_ERROR_ADDRESS_ALREADY_PRESENT doesn't make sense for the UWBS to ever send back if the proper way to update the list of controlees is, as we do here, to send the full list of controlees every time. This means one of two things:
1. STATUS_ERROR_ADDRESS_ALREADY_PRESENT is never sent back as a status when updating controlee list.
2. We misunderstood the proper way to update the list of controlees, and need to redesign this function

Reply:
Those are good observations. Re-reading UCI spec now, I think it's clear what the expectation is:

* To update controlee list outside of an active session, use `SESSION_SET_APP_CONFIG_CMD`. It sets the complete list every time. Since the session isn't active, this makes sense. This can only return 1 of 3 values: `STATUS_INVALID_PARAM`, `STATUS_INVALID_RANGE`, `STATUS_ERROR_SESSION_ACTIVE` (per spec explicitly listing these 3 only as possible `SESSION_SET_APP_CONFIG_RSP::[AppCfgParam[i]]::Status` field values). 

* To update controlee list during an active session, use `SESSION_UPDATE_CONTROLLER_MULTICAST_LIST_CMD`. This has an action argument that says whether the address specfied should be added or deleted. The response to this command can return a bunch of different values, but I believe if the request made it all the way to the device and basic sanity checks passed, it will return `STATUS_OK`. The final results are provided in a `SESSION_UPDATE_CONTROLLER_MULTICAST_LIST_NTF` notification, where the true status of the update is provided, which includes `STATUS_OK_MULTICAST_LIST_UPDATE`, `STATUS_ERROR_MULTICAST_LIST_FULL`, `STATUS_ERROR_KEY_FETCH_FAIL` or `STATUS_ERROR_SUB_SESSION_ID_NOT_FOUND`. This description in the spec if very explicit as it references the table the values come from (Table 27, which are "Multicast update status codes").

Let's keep it simple for now and remove checks on the max number of controlees from this layer, and just return whatever the connector provides as the result. Motivation is that if we make an assumption, let's localize it to one place. Closer to the driver is better, since information from that layer is more accurate.

Reply:
got it. I removed the check on the max number of controlees, but kept the one about the address already present (since it doesn't make sense to send the SESSION_SET_APP_CONFIG_CMD when there is no update).